docs/interfaces.md

```diff
@@ -0,0 +1,568 @@
+# FPC Component Interfaces and States
+
+This document defines the interfaces exposed by the FPC components and their internal state.
+
+
+# ERCC
+
+## Interface:
+The ERCC interface is implemented through a "normal" chaincode interface. All functions can be called by admin or via cc2cc.
+
+```go
+// chaincode interface (exposed to ecc and clients)
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// functions below are dispatched by Invoke implementation
+// Note: Maybe use protobuf-to-json mapping (https://developers.google.com/protocol-buffers/docs/proto3#json)
+// to map the methods and their arguments into to chaincode
+// invocation form { "Function": "..", "Args": [....]}
+
+// returns a set of credentials registered for a given chaincode id
+func queryCCInfos(chaincode_id string) (chaincode_ek []byte, allCredentials []Credentials) {}
+func queryCCInfo(chaincode_id string, enclave_id string) (chaincode_ek []byte, credentials Credentials) {}
```

Comment:
`queryCCInfo` is used by the FPC validation code.

Reply:
there is a new comment regarding the discovery of endorsing FPC-peers.

Reply:
I agree on the following: having a single call to get an enclave's credentials makes sense for validation.
That said, the `chaincode_ek` is unnecessary in that setting.

Also, back to another comment of mine (that I cannot find), now I see why `chaincode_id` is needed: it's because, in ercc's state,  `namespaces/registered/<enclave_id> -> SignedCCKeyRegistrationMessage` does not contain `cc_params`, which is under `<chaincode_id>/<enclave_id>`. So, by reorganizing ercc's state, we can get rid of the `chaincode_id` -- in fact, `enclave_id` is enough the purpose and, if `chaincode_id` is necessary, it can be found in the `credentials`.

Let me propose a possible improvement, having as objective the reduction of: input and output parameters,
* add `func queryChaincodeEncryptionKey(chaincode_id string) (chaincode_ek []byte) {}`
* `queryCCInfos` becomes `func queryListChaincodeEnclaveCredentials(chaincode_id string) (allCredentials []Credentials) {}`
* `queryCCInfo` becomes `func queryEnclaveCredentials(enclave_id string) (credentials Credentials) {}`

In ERCC state:
* `namespaces/registered/<chaincode_id> -> []enclave_ids`
* remove `namespaces/credentials/<chaincode_id>/<enclave_id>` (hence, no duplicate credentials)
* `namespaces/credentials/<enclave_id> -> Credentials` (ercc can still build a list of credentials from a list of ids)
* `namespaces/key_registration/<chaincode_id> -> SignedCCKeyRegistrationMessage` (the encryption key can be taken directly from here)
* in `CCKeyRegistrationMessage`, `enclave_vk` becomes `enclave_id`, which is used to look up `enclave_vk` for signature verification.

Reply:
Thanks Bruno, I partially integrated your proposed changes.

- Introduced `queryChaincodeEncryptionKey` and removed the key from "old" `queryCCInfo(s)`
- Renamed `queryCCInfo` and `queryCCInfos`
- Add clarification that credentials are not stored twice on ercc state; my representation of how to store the credentials was just misleading and confusing.

Regarding removing the `chaincode_id` from the keys and and adding `namespaces/registered/<chaincode_id> -> []enclave_ids` I'm hesitating. Removing as many parameters as possible from the methods exposed by ERCC is a good. 

However, I see the following problem with the proposed layout:
```
namespaces/registered/<chaincode_id> -> []enclave_ids
namespaces/credentials/<enclave_id> -> Credentials
```

Concurrent registration operation:
Let's imagine concurrent enclave registrations for Chaincode A. Let's say a new FPC chaincode is installed and the orgs start setting up their chaincode enclaves and invoke `registerEnclave` at ERCC. (I ignore the fact that MVP does not support this situation but once we are moving forward this can happen).
The implementation of `registerEnclave` will perform the following:
```go
putState("namespaces/credentials/some_new_enclave_id", somecredentials)
registered := getState("namespaces/registered/some_chaincode_id")
registered = append(registered, some_new_enclave_id)
putState("namespaces/registered/some_chaincode_id", registered)
```
These storage operations may lead to write conflicts of concurrent registerEnclave transactions. Even with a single admin issuing the registration of a bunch of enclaves, the admin would need to wait until a register tx is committed to the ledger before registering the next enclave. By integrating the chaincode_id in the key the example code from above is reduced to:
```go
putState("namespaces/credentials/some_chaincode_id/some_new_enclave_id", somecredentials)
```
and no write conflicts will happen.

Another aspect is performance and clean code:
Using `namespaces/credentials/some_chaincode_id/some_new_enclave_id` allows us to perform a range query with `namespaces/credentials/some_chaincode_id` as prefix to query all registered chaincode enclaves with a single `getStateByRange` call instead of first get the all enclave ids and subsequently another getState operation for each enclave id. Note that for each getState invocation, the chaincode shim contacts the peer.

Reply:
as mentioned on Tuesday, you could have the best of both worlds, though, with following schema?
```
namespaces/registered/<chaincode_id>/<enclave_ids> -> defined
namespaces/credentials/<enclave_id> -> Credentials
```

Reply:
that said, i think we should also not overthinking this: Doesn't have to be perfect, good enough should work also :-)

Reply:
> as mentioned on Tuesday, you could have the best of both worlds, though, with following schema?
> 
> ```
> namespaces/registered/<chaincode_id>/<enclave_ids> -> defined
> namespaces/credentials/<enclave_id> -> Credentials
> ```

I agree

> that said, i think we should also not overthinking this: Doesn't have to be perfect, good enough should work also and what you propose definitely does (in particular as we don't really need any lookup `enclave_id->chaincode_id`, so above discussion is a bit academic ...:-)

I agree. Let's go ahead.